src/Analyzers/CSharp/Analyzers/UseIsNullCheck/CSharpUseIsNullCheckOverIsObject.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseIsNullCheck
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal sealed class CSharpUseIsNullCheckOverIsObject : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        public CSharpUseIsNullCheckOverIsObject()
+            : base(IDEDiagnosticIds.UseIsNullOverIsObjectDiagnosticId,
+                   EnforceOnBuildValues.UseIsNullCheck,
+                   CSharpCodeStyleOptions.PreferIsNullCheckOverIsObject,
+                   CSharpAnalyzersResources.Use_is_null_check,
+                   new LocalizableResourceString(nameof(AnalyzersResources.Null_check_can_be_simplified), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(context =>
+            {
+                var objectType = context.Compilation.ObjectType;
+                if (objectType.TypeKind == TypeKind.Error)
+                {
+                    return;
+                }
+
+                // All trees should have the same language version. Bail-out early in compilation start instead of checking every tree.
+                var tree = context.Compilation.SyntaxTrees.FirstOrDefault();
+                if (tree is null || ((CSharpParseOptions)tree.Options).LanguageVersion < LanguageVersion.CSharp9)
+                {
+                    return;
+                }
+
+                context.RegisterOperationAction(c => AnalyzeOperation(c, objectType), OperationKind.IsType, OperationKind.NegatedPattern);
+            });
+        }
+
+        private void AnalyzeOperation(OperationAnalysisContext context, INamedTypeSymbol objectType)
+        {
+            var option = context.Options.GetOption(CSharpCodeStyleOptions.PreferIsNullCheckOverIsObject, context.Operation.Syntax.SyntaxTree, context.CancellationToken);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            if (ShouldReportDiagnostic(context.Operation, objectType))
+            {
+                var severity = option.Notification.Severity;
+                context.ReportDiagnostic(
+                    DiagnosticHelper.Create(
+                        Descriptor, context.Operation.Syntax.GetLocation(), severity, additionalLocations: null, properties: null));
```

Comment:
i'd like for us to ensure the syntax is what we expect as well.  in case this pattern was somehow emitted in the bound nodes but on syntax we don't expect.

Reply:
@CyrusNajmabadi I'm checking the syntax in the codefix and throwing `ExceptionUtilities.UnexpectedValue` on unexpected syntax.

Reply:
Definitely not the way to do it :-)
Analyzers should not report issues if the fix would crash on them :-)

Reply:
You sound cleanly bail out in release mode. We can consider a debug assert for debug mode, but my preference is not too have that. 

We can effectively drive completeness in these cases through dogfooding and user reports.